include/utils/str.h

```diff
@@ -3,44 +3,196 @@
 
 #include <stddef.h>
 #include <stdarg.h>
+#include <stdbool.h>
 
-typedef struct str_t {
+typedef struct {
     size_t len;
     char* data;
 } str;
 
-void str_create(str *string);
-void str_create_from_cstr(str *string, const char *cstr);
-void str_create_from_data(str *string, const char *data, size_t len);
+/**
+ * @brief Create an empty string object.
+ * @details Creates a new, empty string object. After creating the object, it must be freed
+ *          using the str_free. 
+ * @param dst Target string object
+ */
+void str_create(str *dst);
+
+/**
+ * @brief Create a new string from existing string.
+ * @details Source string contents will be copied.
+ * @param dst Target string object
+ * @param src Source string to copy
+ */
+void str_from(str *dst, const str *src);
+
+/**
+ * @brief Create a string object from C string (null terminated buffer).
+ * @details Source string contents will be copied.
+ * @param dst Target string object
+ * @param src Source C string to copy
+ */
+void str_from_c(str *dst, const char *src);
+
+/**
+ * @brief Create a string object from memory buffer.
+ * @details Source buffer content will be copied. Please make sure that the source
+ *          buffer contains no null characters, and is actual printable text string.
+ * @param dst Target string object
+ * @param buf Source buffer object
+ * @param len Source buffer object length (bytes)
+ */
+void str_from_buf(str *dst, const char *buf, size_t len);
+
+/**
+ * @brief Create a string object by format string
+ * @details This uses snprintf internally, so usual printf format strings are valid.
+ * @param dst Target string object
+ * @param format Format string
+ * @param ... Arguments list
+ */
+void str_from_format(str *dst, const char *format, ...);
+
+/**
+ * @brief Create a string object from range slice
+ * @details This will copy a text range from another string object. Note that source string
+ *          must not be same as target string!
+ * @param dst Target string object
+ * @param src Source string object
+ * @param start Slice start position
+ * @param end Slice end position
+ */
+void str_from_slice(str *dst, const str *src, size_t start, size_t end);
+
+/**
+ * @brief Free string object
+ * @details Frees up any memory used by the string object. Usage of the string after freeing it
+ *          is undefined behaviour.
+ * 
+ * @param string String to free.
```

Comment:
This now had period at the end, but others don't

Reply:
asdf.